### PR TITLE
Add interface_capabilities helper

### PR DESCRIPTION
**Pull request by chrisvanheuveln**
* New Type, Provider, and beaker utility; for use with interface testing.
* Tested mainly on n7k.
* Considered using facter for this but the overhead in processing time was more than desired and I didn't want to penalize every single puppet run; therefore it made more sense to make this an on-demand utility since it's only going to be used by beaker.
* Manual test:

```
    cap = interface_capabilities(agent, 'ethernet9/1')
    puts 'Speed: ' + cap['Speed']
    puts 'Duplex: ' + cap['Duplex']
```

Result:

```
Speed: 10000,40000
Duplex: full
```

### CODE REVIEW

File: lib/puppet/provider/cisco_interface_capabilities/cisco.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+end
+
+Puppet::Type.type(:cisco_interface_capabilities).provide(:cisco) do
+  confine feature: :cisco_node_utils
+  defaultfor operatingsystem: [:ios_xr, :nexus]
+
+  def self.instances
+    interfaces = []
+    Cisco::Interface.interfaces.each do |intf, _|
+      next if intf.match(/mgmt/i)
```
Comment (robert-w-gries) [inline lib/puppet/provider/cisco_interface_capabilities/cisco.rb]: Make sure to exclude loopback/port_channel interfaces as well

```diff
-      next if intf.match(/mgmt/i)
+      next unless intf.match(/ethernet/i)
```

File: tests/beaker_tests/lib/utilitylib.rb [hunk 2]
```diff
@@ -1029,3 +1029,39 @@ def find_interface(tests, id=nil, skipcheck=true)
...
+
+  str = stdout[/\[(.*)\]/][1..-2]
+  return {} if str.nil?
+
+  # 'Model: N7K-F312FQ-25', '', 'Type (SFP capable):    QSFP-40G-4SFP10G', '',
+  # 'Speed: 10000,40000', '', 'Duplex: full', ''
+  str.delete!("'")
+
+  # Model: N7K-F312FQ-25, , Type (SFP capable):    QSFP-40G-4SFP10G, ,
+  # Speed: 10000,40000, , Duplex: full, ,
```
Comment (robert-w-gries) [inline tests/beaker_tests/lib/utilitylib.rb]: I thought `str.delete!("'")` would delete the empty entries?

Comment (robert-w-gries) [inline tests/beaker_tests/lib/utilitylib.rb]: Ah nevermind, I was thinking of the similar array.delete from `node_utils`

**Comment by robert-w-gries** [comment]
Code looks good.  But I feel a bit uneasy about adding a provider without adding a corresponding test

File: lib/puppet/provider/cisco_interface_capabilities/cisco.rb (hunk 1 above)
Comment (chrisvanheuveln) [inline lib/puppet/provider/cisco_interface_capabilities/cisco.rb]: fixed

**Comment by chrisvanheuveln** [comment]
okay, give me a sec...

File: tests/beaker_tests/lib/utilitylib.rb [hunk 3]
```diff
@@ -1029,3 +1029,39 @@ def find_interface(tests, id=nil, skipcheck=true)
   end
   intf
 end
+
+# Use puppet resource to get interface capability information.
+# TBD: Facter may be a better home for this method but the performance hit
+# appears to be 2s per hundred interfaces so it works better for now as an
+# on-demand method.
+def interface_capabilities(agent, intf)
+  cmd = get_namespace_cmd(agent, PUPPET_BINPATH +
```
Comment (mikewiebe) [inline tests/beaker_tests/lib/utilitylib.rb]: Not sure how you want to handle this, but after https://github.com/cisco/cisco-network-puppet-module/pull/233 gets merged I believe you no longer need to call `get_namespace_cmd`.

File: lib/puppet/provider/cisco_interface_capabilities/cisco.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  require 'puppet_x/cisco/autogen'
+rescue LoadError # seen on master, not on agent
+  # See longstanding Puppet issues #4248, #7316, #14073, #14149, etc. Ugh.
+  require File.expand_path(File.join(File.dirname(__FILE__), '..', '..', '..',
+                                     'puppet_x', 'cisco', 'autogen.rb'))
+end
+
+Puppet::Type.type(:cisco_interface_capabilities).provide(:cisco) do
+  confine feature: :cisco_node_utils
+  defaultfor operatingsystem: [:ios_xr, :nexus]
```
Comment (mikewiebe) [inline lib/puppet/provider/cisco_interface_capabilities/cisco.rb]: Is this command supported on XR?

Comment (chrisvanheuveln) [inline lib/puppet/provider/cisco_interface_capabilities/cisco.rb]: No. But this is not the place for platform / OS checks.

File: tests/beaker_tests/lib/utilitylib.rb (hunk 3 above)
Comment (chrisvanheuveln) [inline tests/beaker_tests/lib/utilitylib.rb]: I know I don't want to change it until *after* it gets merged. :-)

File: lib/puppet/provider/cisco_interface_capabilities/cisco.rb (hunk 4 above)
Comment (mikewiebe) [inline lib/puppet/provider/cisco_interface_capabilities/cisco.rb]: Agreed, but you probably don't need `:ios_xr` in your `defaultfor`

**Comment by mikewiebe** [comment]
:+1:

**Pull request merged**